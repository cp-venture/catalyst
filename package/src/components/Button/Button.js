import React from "react";
import PropTypes from "prop-types";
import CircularProgress from "@material-ui/core/CircularProgress";
import MuiButton from "@material-ui/core/Button";
import makeStyles from "@material-ui/core/styles/makeStyles";

const useStyles = makeStyles((theme) => ({
  buttonProgress: {
    marginLeft: theme.spacing.unit
  },
  containedPrimary: {
    "color": theme.palette.primary.contrastText,
    "backgroundColor": theme.palette.colors.red,
    "&:hover": {
      "backgroundColor": theme.palette.colors.redHover,
      // Reset on touch devices, it doesn't add specificity
      "@media (hover: none)": {
        backgroundColor: theme.palette.colors.redHover
      }
    }
  },
  outlinedPrimary: {
    "color": theme.palette.colors.red,
    "border": `1px solid ${theme.palette.colors.red}`,
    "&:hover": {
      "border": `1px solid ${theme.palette.colors.redBorder}`,
      "backgroundColor": theme.palette.colors.redBackground,
      // Reset on touch devices, it doesn't add specificity
      "@media (hover: none)": {
        backgroundColor: "transparent"
      }
    }
  }
}));
/**
 * @name Button
 * @param {Object} props Component props
 * @returns {React.Component} returns a React component
 */
function Button(props) {
  const {
    children,
    color,
    disabled,
    isWaiting,
    ...otherProps
  } = props;

  const classes = useStyles();

  if (color === "danger") {
    return (
      <MuiButton
        classes={{
          containedPrimary: classes.containedPrimary,
          outlinedPrimary: classes.outlinedPrimary
        }}
        color="primary"
        disabled={disabled || isWaiting}
        {...otherProps}
      >
        {children}
        {isWaiting && <CircularProgress size={16} className={classes.buttonProgress} />}
      </MuiButton>
    );
  }

  return (
    <MuiButton
      color={color}
      disabled={disabled || isWaiting}
      {...otherProps}
    >
      {children}
      {isWaiting && <CircularProgress size={16} className={classes.buttonProgress} />}
    </MuiButton>
  );
}

Button.propTypes = {
  /**
   * What goes here?
   */
  children: PropTypes.node,
  /**
   * CSS Classes
   */
  classes: PropTypes.object,
  /**
   * Options: `default` | `inherit` | `primary` | `secondary` | `danger`
   */
  color: PropTypes.enum,
  /**
   * If `true`, the button will be disabled.
   */
  disabled: PropTypes.bool, // eslint-disable-line
  /**
   * If `true`, the CircularProgress will be displayed.
   */
  isWaiting: PropTypes.bool,
  /**
   * onClick callback
   */
  onClick: PropTypes.func
};

export default Button;
