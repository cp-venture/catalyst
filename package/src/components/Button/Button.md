### Overview

The Catalyst Button inherits from the Material-UI [Button component](https://material-ui.com/components/buttons/). Refer to tne Material-UI [Button API docs](https://material-ui.com/api/button/) for more information. 

### Usage


#### Default Catalyst button

This is what a button with all the default prop options looks like:

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button>Text-only</Button>
  </div>
</div>
```

It's a button with `variant` set to `text`, `size` set to `medium`, `color` set to `default`.

#### Material UI options

- **Solid button**: The solid button is used for a primary action in a modal, card, large view and generally throughout.
- **Outline button**: The outline button is used for a secondary or dismissive action. The outline button should be paired with the solid button in cases such as dismissing a modal or canceling an action.

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button>Text-only</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="contained">Solid</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="outlined">Outlined</Button>
  </div>
</div>
```

- **Disabled**: Docs go here. When do you use this button in Catalyst?

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="contained" isWaiting>isWaiting</Button>
  </div>
</div>
```

- **Full Width**: Docs go here. When do you use this button in Catalyst?

```jsx
<div style={{ display: "block" }}>
  <Button variant="contained" fullWidth>FullWidth</Button>
</div>
```

#### Catalyst-custom buttons

- **Danger button**: The danger button is used for a destructive action that is difficult to recover from such as deleting information. The danger button should be used at the point that the destructive action actually takes place. For example, you can have a delete button as a secondary action on a page and in this case you would use a an outline button, the outline button would then trigger a modal confirmation, which is where you would use the danger button.

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="contained" color="danger">Danger - Contained</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="outlined" color="danger">Danger - Outlined</Button>
  </div>
</div>
```

- **isWaiting**: Docs go here.

```jsx
<div style={{ display: "flex" }}>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="contained" isWaiting>isWaiting - Contained</Button>
  </div>
  <div style={{ marginRight: "1rem" }}>
    <Button variant="outlined" isWaiting>isWaiting - Outlined</Button>
  </div>
</div>
```

