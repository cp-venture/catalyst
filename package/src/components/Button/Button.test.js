import React from "react";
import { render } from "../../tests/index.js";
import Button from "./Button";

test("basic snapshot - only default props", async () => {
  const { asFragment } = await render(<Button className="myBtn">Submit</Button>);
  expect(asFragment()).toMatchSnapshot();
});

// test("basic snapshot - with props for an Error button", async () => {
//   const { asFragment } = render(<Button className="myBtn" variant="contained" color="danger">Submit</Button>);
//   expect(asFragment()).toMatchSnapshot();
// });
