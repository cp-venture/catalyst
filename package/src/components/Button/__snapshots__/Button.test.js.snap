// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root a b MuiButton-text"
  components={
    Object {
      "Button": [Function],
    }
  }
  disabled={false}
  href={null}
  mini={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  title="title"
  type="button"
>
  <span
    className="MuiButton-label"
  >
    Submit
  </span>
</button>
`;
